lib/index.js

```diff
@@ -308,167 +307,123 @@ KnexMigrator.prototype.migrate = function migrate(options) {
+            if (onlyVersion && version !== onlyVersion) {
+                debug('Do not execute: ' + version);
                 return;
             }
+        });
 
-            return Promise.each(versionsToMigrate, function (versionToMigrate) {
-                return self._migrateTo({
-                    version: versionToMigrate,
-                    only: onlyFile,
-                    hooks: hooks
-                });
+        if (onlyVersion) {
+            // CASE: filter out versions which should not run
+            let containsVersion = _.find(result, function (_obj, key) {
+                return key === onlyVersion;
             });
-        })
-        .then(function executeAfterHook() {
-            if (!versionsToMigrate.length) {
-                return;
-            }
 
-            if (hooks.after) {
-                debug('After hook');
-                return hooks.after({
-                    connection: self.connection
-                });
-            }
-        })
-        .then(function () {
-            return locking.unlock(self.connection);
-        })
-        .catch(function (err) {
-            // CASE: Do not rollback if migrations are locked
-            if (err instanceof errors.MigrationsAreLockedError) {
-                throw err;
+            if (!containsVersion) {
+                logging.warn('Cannot find requested version: ' + onlyVersion);
             }
+        }
 
-            // CASE: Do not rollback migration scripts, if lock error
-            if (err instanceof errors.LockError) {
-                throw err;
+        _.each(result, function (value, version) {
+            // CASE: compare files on disk with files in database
+            if (value.expected !== value.actual) {
+                debug('Need to execute migrations for: ' + version);
+                versionsToMigrate.push(version);
             }
+        });
 
-            // CASE: ETIMEDOUT, ENOTFOUND
-            if (err instanceof errors.DatabaseError) {
-                throw err;
+        if (versionsToMigrate.length) {
+            if (hooks.before) {
+                debug('Before hook');
+                await hooks.before({
+                    connection: this.connection
+                });
             }
 
-            if (err.context && err.context.name) {
-                debug(`Task failed: ${err.context.name}`);
-            }
+            for (const versionToMigrate of versionsToMigrate) {
+                try {
+                    await this._migrateTo({
+                        version: versionToMigrate,
+                        only: onlyFile,
+                        hooks: hooks
+                    });
+                } catch (err) {
+                    if (err.context && err.context.name) {
+                        debug(`Task failed: ${err.context.name}`);
+                    }
 
-            debug(`Rolling back: ${err.message}`);
+                    debug(`Rolling back: ${err.message}`);
```

Comment:
Perhaps we could change this to `logging.warn/error`? When someone updates across several versions and tables/columns are being added and removed, it can be a little difficult to figure out where in the logs we're starting to roll back

Reply:
Yeah that's a fine idea - will do that now! Any other logs you'd like to see whilst I'm there?

Reply:
Two things:

1. Can we log when we're just about to start migrations? Otherwise we go straight from the DB backup into whatever the migrations themselves log and there's no "hey we're about to drop down into a migration and start running it".

```
[2021-11-02 13:55:14] INFO Creating database backup
[2021-11-02 13:55:14] INFO Database backup written to: /home/xxx.json
[I feel there should be something here like "Running migrations"]
[2021-11-02 13:55:14] INFO Dropping table: offer_redemptions
```

2. I've noticed that when we do an explicit rollback (ie. `knex-migrator rollback --v 4.12 --force`) we print the following at the end. Perhaps we can also log that here, and _then_ print the error?
```
[2021-11-02 13:51:45] INFO Rollback was successful.
```

This would really help when we get bug reports and try to decipher whether or not it was a migration or the rollback that failed.

The key thing is to make it obvious that we 1) started to migrate, 2) hit an error and need to rollback, 3) rolled back and 4) what the full error is 🙂

Reply:
Okay that all makes sense - will sort it!

Reply:
How's this look?
```
❯ yarn run knex-migrator migrate --v 4.22 --force 
yarn run v1.22.5
warning ghost@4.20.4: The engine "cli" appears to be invalid.
$ /home/egg/usr/src/ghost.org/Ghost-migrations/node_modules/.bin/knex-migrator migrate --v 4.22 --force
[2021-11-02 15:28:51] INFO Running migrations.
[2021-11-02 14:28:51] INFO Creating database backup
[2021-11-02 14:28:51] INFO Database backup written to: /home/egg/usr/src/ghost.org/Ghost-migrations/content/data/ghost.2021-11-02-14-28-51.json
[2021-11-02 14:28:51] INFO 4.21 First up
[2021-11-02 14:28:51] INFO 4.21 Second up
[2021-11-02 14:28:51] INFO 4.21 Third up
[2021-11-02 14:28:51] INFO 4.22 First up
[2021-11-02 14:28:51] INFO 4.22 Second up - bout to throw
[2021-11-02 15:28:51] INFO Rolling back: fuck.
[2021-11-02 14:28:51] INFO 4.22 First down
[2021-11-02 14:28:51] INFO 4.21 Third down
[2021-11-02 14:28:51] INFO 4.21 Second down
[2021-11-02 14:28:51] INFO 4.21 First down
[2021-11-02 15:28:51] INFO Rollback was successful.
[2021-11-02 15:28:51] ERROR fuck

fuck

{"config":{"transaction":true},"name":"4.22-02-second.js"}
"Error occurred while executing the following migration: 4.22-02-second.js"

Error ID:
    300

----------------------------------------

MigrationScriptError: fuck
    at MigrationScriptError.KnexMigrateError (/home/egg/usr/src/ghost.org/knex-migrator/lib/errors.js:7:26)
    at new MigrationScriptError (/home/egg/usr/src/ghost.org/knex-migrator/lib/errors.js:25:26)
    at /home/egg/usr/src/ghost.org/knex-migrator/lib/index.js:1028:19

Error: fuck
    at up (/home/egg/usr/src/ghost.org/Ghost-migrations/core/server/data/migrations/versions/4.22/4.22-02-second.js:7:15)
    at Object.up (/home/egg/usr/src/ghost.org/Ghost-migrations/core/server/data/migrations/utils.js:308:19)
    at /home/egg/usr/src/ghost.org/knex-migrator/lib/index.js:978:33
    at /home/egg/usr/src/ghost.org/knex-migrator/node_modules/knex/lib/transaction.js:183:22

error Command failed with exit code 1.
info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.
```

Reply:
Can I switch the order around to 

```
[2021-11-02 14:28:51] INFO Creating database backup
[2021-11-02 14:28:51] INFO Database backup written to: /home/egg/usr/src/ghost.org/Ghost-migrations/content/data/ghost.2021-11-02-14-28-51.json
[2021-11-02 15:28:51] INFO Running migrations.
[2021-11-02 14:28:51] INFO 4.21 First up
```

so it logs "Running migrations." after the `before` hook

Reply:
Yup!!